### PR TITLE
Expose Hamiltonian Generation

### PR DESCRIPTION
**Pull request by johnzl-777**
This allows users to obtain the raw Hamiltonian generated by the program

### CODE REVIEW

File: src/bloqade/ir/routine/bloqade.py [hunk 1]
```diff
@@ -362,3 +419,22 @@ def run_callback(
...
         return results
+
+    def hamiltonian(
+        self,
+        *args: LiteralType,
+        blockade_radius: float = 0.0,
+        waveform_runtime: str = "interpret",
+        cache_matrices: bool = False,
+    ) -> Iterator["HamiltonianData"]:
+        pass
```
Comment (weinbe58) [inline src/bloqade/ir/routine/bloqade.py:430]:
```suggestion
```

**Review by weinbe58 — CHANGES_REQUESTED** [review]
Need to add some kind of test, maybe just make sure the hamiltonian for a single atom is correct.

File: src/bloqade/ir/routine/bloqade.py [hunk 2]
```diff
@@ -19,6 +30,44 @@ def python(self):
...
+    """Data class to hold the program ir and metadata for a given set of parameters"""
+
+    task_id: int
+    emulator_ir: EmulatorProgram
+    metadata_dict: Dict[str, LiteralType]
+
+    @property
+    def metadata(self) -> NamedTuple:
+        MetaData = namedtuple("MetaData", self.metadata_dict.keys())
+        return MetaData(**{k: float(v) for k, v in self.metadata_dict.items()})
```
Comment (weinbe58) [inline src/bloqade/ir/routine/bloqade.py:44]: still not called here

**Comment by johnzl-777** [comment]
@weinbe58 I realize that with the new API I could go around and refactor all the tests in `test_hamiltonian_codege.py` to use it. 

I made a single test based off your `test_2_level_uniform_detuning` unit test for now but if you'd like me to proceed with the refactor after taking a look at what I've added, just let me know!

**Pull request merged**